### ISSUE TITLE
Add a "Preview Bake" button for BakedLightmap for quick preview bakes

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot-proposals#2029]
### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

With https://github.com/godotengine/godot/pull/44628 and in the `master` branch, BakedLightmap can make great-looking scenes, but the bake time is too long to quickly iterate on. 

The user can temporarily switch to lower quality settings, but this is cumbersome and you risk accidentally committing the new settings to version control. This can cause difficult-to-diagnose issues down the line.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

When baking lightmaps, allow using low-quality settings in a *temporary* manner that doesn't affect the BakedLightmap's quality properties.

This kind of feature has been available for decades in various 3D modellers and map editors such as GtkRadiant, so there's a fair amount of track record for it.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Have two buttons instead of one for baking:

- Preview Bake. This bakes using low-quality settings defined in the editor settings (or perhaps project settings, for consistency across project members).
- "Final" Bake (name is subject to change). This uses the quality settings defined in the BakedLightmap node.

If the user has performed a preview bake, there should be some kind of watermark on the 3D viewport to inform the user when the BakedLightmap node is selected. This is to make sure people don't ship preview bakes to production :slightly_smiling_face: 

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This can likely be implemented in an add-on, but it's not exactly trivial.

### Is there a reason why this should be core and not an add-on in the asset library?

This is about improving the user experience by decreasing iteration times, so I'd argue it's important to have this in core.

**Comment by lander-vr** [comment]
I wasn't 100% sure whether to reply here or on https://github.com/godotengine/godot/pull/80518, since it's a bit about both the proposal and the PR.

The biggest time sink for artists when using lightmapped GI is validating the lightmap setup when baking on final quality, especially in larger levels this can be a huge pain as doing clean high quality lightbakes can often mean hour(s) of baketime, and since the visual result of lightmaps can change (often unexpectedly) between quality settings, the only way to validate if your final lightmap works is by doing a final quality lightmap bake. This is something that I'd expect a preview option to help with, but https://github.com/godotengine/godot/pull/80518 essentially does a low quality bake and I'm a little doubtful if that's very helpful to artists.

I think a better alternative functionality for preview bakes (though a little more complex than the existing PR) is baking a visibility mask doing traces from pixels to the camera as an initialization step, hits are black (occluded by geometry), outside frustum is black, dilate it a few pixels (denoiser range + maybe a few pixels extra for a bit of leeway) and then feed that mask into the lightmapper and denoiser to skip traces from pixels that are masked out.
This way we can keep the users current quality settings but drastically reduce the baketime, and allow the user to get a preview bake that's representative of the settings they are adjusting without being forced to bake the entire scene.

I do really like the functionality of being able to do a bake without being forced to commit to it, and I'm thinking this could be an interesting default behavior since sometimes a lightbake doesn't turn out as you expect, and the previous version is preferred. Maybe the previous lightmap could remain saved instead of being immediately discarded, so you can always roll back? (e.g. how in Blender, when you save a file your previous save turns into a .blend1 file so you can always roll back one version)

**Comment by passivestar** [comment]
I think both are useful. Camera bake for quick final quality preview and full scene low-res bake for when you're doing level blockouts. I prefer walking around lightmapped blockouts because lightmaps change the feel of the space drastically

**Comment by lander-vr** [comment]
> full scene low-res bake for when you're doing level blockouts

You wouldn't just bake at a low quality preset in that case? Especially since the preview option removes any control over the lightmap options, I can imagine that the "locked in" preview bake settings in some cases can be inadequate even for a preview (high noise bakes, using emissive surfaces, indoor environments with only indirect lighting, ...).

I'm not convinced that doing that requires a separate feature when you can just bake on the low quality preset and set your texel scale on `0.5` for a full scene preview, and retain control over the settings so the user can decide on the quality vs speed tradeoff themselves.

**Comment by passivestar** [comment]
@lander-vr  This proposal implies low quality settings are user-defined, so you have full control over them:
> Preview Bake. This bakes using low-quality settings defined in the editor settings (or perhaps project settings, for consistency across project members)

It's quite a bit more convenient than to manually change multiple settings back and forth between bakes when i.e you're done grayboxing/iterating and want to test the map with full-res lightmaps in all its glory. I have separate preview bake buttons in a custom baking addon I made and it saves me time:

<img width="318" height="272" alt="Image" src="https://github.com/user-attachments/assets/27840ea0-9385-465e-b7b0-4833607f40f8" />

**Comment by lander-vr** [comment]
>  This proposal implies low quality settings are user-defined, so you have full control over them.

Ah yes, this makes sense, I missed that!

> It's quite a bit more convenient than to manually change multiple settings back and forth between bakes when i.e you're done grayboxing/iterating and want to test the map with full-res lightmaps in all its glory.

Totally agreed, but I think with that in mind it makes more sense for the quality settings to be an overarching preset in the lightmapGI node, instead of just being representative of the ray count. Meaning that all lightmapGI node settings, like supersampling, bounces, ray count, etc.., are tied to a specific quality preset. Then a preview quality mode could nicely fit inside the lightmapgi node and is just a quality setting you can switch to, without losing your custom settings on another quality preset.

I'm not a fan of placing variables like this in the project settings, it feels a little arbitrary to me.

I'm not sure if that's possible though, to "tuck away" user changed settings and bring them back later - so maybe the PR as is is the most straightforward approach.

**Comment by passivestar** [comment]
I think project settings driven button is better than a preset because:

1. A bit less noise in VC because you don't change the node settings
2. Low quality doesn't mean preview. Low quality can mean shippable for low-end devices if configured properly. Preview means not for shipping 
3.  Scales better for when godot gets some kind of a bake manager to bake multiple lightmaps at once. Baking multiple previews will be easier when preview logic is not in the node settings

**Comment by lander-vr** [comment]
> Baking multiple previews will be easier when preview logic is not in the node settings

Yes this makes a lot of sense, that's a great argument for this to be in project settings.
Thanks for giving some context on why this implementation makes sense as is, I'm on board with this now.

And I suppose for frustum only baking we can just add another bake button.
